File: kaantelypeli/src/main/java/kaantelypeli/utils/FXUtils.java
package kaantelypeli.utils;

import javafx.event.ActionEvent;
import javafx.event.EventHandler;
import javafx.scene.control.Button;
import javafx.scene.control.ChoiceDialog;
import javafx.stage.Stage;
import kong.unirest.Unirest;

import java.util.List;

public class FXUtils {
    private FXUtils() {
        throw new IllegalStateException("Utility class");
    }

    /**
     * Create a JavaFX button
     * @param name Name and Id for the button
     * @param action Action to call on button press
     * @return Button with the given options
     */
    public static Button button(String name, EventHandler<ActionEvent> action) {
        Button button = new Button(name);
        button.setOnAction(action);
        button.setId(name);
        return button;
    }

    /**
     * Create a list selector for a JSON list retrieved from URL
     * @param url list URL
     * @param stage stage to parent this choice dialog
     * @return ChoiceDialog with the given options
     */
    public static ChoiceDialog<String> selector(String url, Stage stage) {
        ChoiceDialog<String> choice = new ChoiceDialog<>();
        choice.initOwner(stage);
        choice.setHeaderText("Level to load");
        choice.setGraphic(null);
        choice.setTitle(null);
        List<Object> options = Unirest.get(url).asJson().getBody().getArray().toList();
        for (Object i : options) {
            choice.getItems().add(url + '/' + i);
        }
        return choice;
    }
}

File: kaantelypeli/src/main/java/kaantelypeli/utils/FXUtils.kt
package kaantelypeli.utils

import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.scene.control.Button
import javafx.scene.control.ChoiceDialog
import javafx.scene.image.Image
import javafx.scene.paint.ImagePattern
import javafx.scene.shape.Rectangle
import javafx.stage.Stage
import kong.unirest.Unirest
import org.tinylog.Logger

/**
 * Create a JavaFX button.
 * @param name Name and Id for the button
 * @param action Action to call on button press
 * @return Button with the given options
 */
fun createButton(
    name: String?,
    callback: (ActionEvent) -> Unit
): Button {
    val button = Button(name)
    val action = EventHandler<ActionEvent> { event -> callback(event) }
    button.onAction = action
    button.id = name
    return button
}

/**
 * Create a list selector for a JSON list retrieved from URL.
 * @param url list URL
 * @param stage stage to parent this choice dialog
 * @return ChoiceDialog with the given options
 */
fun selector(url: String, stage: Stage?): ChoiceDialog<String> {
    val choice = ChoiceDialog<String>()
    choice.initOwner(stage)
    choice.headerText = "Level to load"
    choice.graphic = null
    choice.title = null
    Logger.trace("GETing: {}", url)
    val options = Unirest.get(url).asJson().body.array.toList()
    for (i in options) {
        choice.items.add(i.toString())
    }
    return choice
}

fun IntRectangle(i: Int, i1: Int, i2: Int, i3: Int): Rectangle {
    return Rectangle(i.toDouble(), i1.toDouble(), i2.toDouble(), i3.toDouble())
}

fun IntImagePattern(sprite: Image, i: Int, i1: Int, i2: Int, i3: Int, b: Boolean): ImagePattern {
    return ImagePattern(sprite, i.toDouble(), i1.toDouble(), i2.toDouble(), i3.toDouble(), b)

}
